### PR TITLE
Update onMouseMoveSecondary to allow offset adjustments

### PR DESCRIPTION
**Pull request by spurin**
Fixes: #4245

Blocked by: #7522, #7569

---

We now have a working prototype where onMouseMoveSecondary can be adjusted via an environment variable at startup. This implementation is server-side and relates to the issue discussed at #4245 (credit to @neworld for their initial implementation).

To double the mouse speed, use the following on the server side:

```
SYNERGY_MOUSE_ADJUSTMENT=2; open /Applications/Synergy.app
```

This results in consistent mouse speed across my Mac OS X server and HiDPI Windows client, achieving the desired functional outcome.

To reduce the speed, you can modify the value like so:

```
SYNERGY_MOUSE_ADJUSTMENT=0.3; open /Applications/Synergy.app
```

In the future, this could ideally be managed through the UI, with the ability to apply different speed adjustments for specific targets, rather than having a global setting. In my case, I only have two systems, so this works fine, but others may require more granular control.

Consider this a proof of concept for reference, rather than code intended for direct inclusion in the upstream codebase.

### CODE REVIEW

**Comment by nbolton** [comment]
@spurin Oh! `getenv`! Great idea. Very mergeable and easier than fiddling with CLI args. I'll test in the morning 👍

**Comment by spurin** [comment]
Regarding the setting of 2x that I'm using, the mouse speed is perfect and consistent now.

I wonder on the explanation for this.

I have 2 x 4K screens, 3840x2160

Mac Retina is a 1080p "like" resolution on a 4K screen, providing that retina sharpness and crispness by pixel doubling.

My windows desktop is 4K, albeit with a percentage zoom in.

I believe that the retina aspect on mac may be treating the pixel count as a 1080p screen, when we switch to windows, the pixel count is double, hence the slow down of the mouse movement.

Coincidently, when I tried Share Mouse, I also needed to move the bar to 200% to get the equivalent result.

Mentioning this as retina mode is standard on Mac OS and it may be worthwhile thinking about this from a UI/settings implementation.

**Comment by neworld** [comment]
As the user, I can recommend rounding fractional numbers to the `1` because having a 0.5 coefficient means the mouse won't move during 1px movements. It wasn't enjoyable. Because of that, I [implemented flooring](https://github.com/neworld/personal-aur-synergy/blob/master/trunk/adjust_low_dpi_mouse_speed.patch#L11-L12) (instead of rounding) with my limited knowledge of C++.

**Comment by nbolton** [comment]
@spurin Heads up, we're renaming the upstream 'Synergy Community Edition' to Deskflow. Once that's done we can work on getting this PR merged. Sorry for the delay. Discussion: https://github.com/deskflow/deskflow/discussions/7517

**Comment by spurin** [comment]
Thanks for the heads up, interesting name. I like it!

**Comment by sithlord48** [comment]
Would it be possible to make this part of the client config ?

**Comment by nbolton** [comment]
> Would it be possible to make this part of the client/server config ?

I think so. Env var is good for developer testing, but for regular users, config would be easier.

**Comment by nbolton** [comment]
@spurin Would you like us to fix the [SonarCloud-idenfied smells](https://sonarcloud.io/project/issues?issueStatuses=OPEN%2CCONFIRMED&pullRequest=7511&types=CODE_SMELL&inNewCodePeriod=true&sinceLeakPeriod=true&id=deskflow_deskflow) or would you prefer we do it?

**Comment by spurin** [comment]
@nbolton I'm more than happy for yourself or anybody to take this on and use/re-use/update the code as required to make it useful for the task in hand.

For context, I treated this as a functional prototype and for my own personal needs at present it suits me well.

Whilst I'd rather have the slick features of Synergy 3, the mouse aspect is currently more important at present (am doing a lot of design/creative work between mac and windows). So I'm happy to live with this workaround until it takes a better form in the future.

You steer your own ship much better than I, appreciate the ask. In this case please do as you see fit with it 👍

**Comment by nbolton** [comment]
> I'm more than happy for yourself or anybody to take this on

No problem, I'll take a look at getting these issues fixed soon so we can merge this in. I appreciate the time you've spent so far and thanks for sharing your solution with us.

**Comment by sithlord48** [comment]
If we land this without making the config / gui changes we need to be sure this gets documented so we do not forget todo it .

**Comment by sithlord48** [comment]
@nbolton  iirc we were testing our linting on this pr ? that is all im waiting for to land

**Comment by nbolton** [comment]
> that is all im waiting for to land

3 things:
[ ] Testing of Clang lint workflow -- though we don't necessarily need this to block, but it would be a nice test case
[ ] [Code smells](https://sonarcloud.io/project/issues?issueStatuses=OPEN%2CCONFIRMED&sinceLeakPeriod=true&pullRequest=7511&id=deskflow_deskflow) (maintainability/clean code) need fixing (picked up by SonarCloud) -- I plan on fixing these after I'm finished fiddling with CI
[ ] Need to raise an issue about adding a GUI setting for the scroll tweak values

**Comment by nbolton** [comment]
> As the user, I can recommend rounding fractional numbers to the `1` because having a 0.5 coefficient means the mouse won't move during 1px movements. It wasn't enjoyable. Because of that, I [implemented flooring](https://github.com/neworld/personal-aur-synergy/blob/master/trunk/adjust_low_dpi_mouse_speed.patch#L11-L12) (instead of rounding) with my limited knowledge of C++.

@neworld Do you think the current PR needs any more changes?

@spurin Are you happy for this to be merged in its current state? I only applied the code style.

**Comment by spurin** [comment]
> @spurin Are you happy for this to be merged in its current state?

All good by me @nbolton

File: src/lib/server/Server.cpp [hunk 1]
```diff
@@ -1790,7 +1791,31 @@ void Server::sendDragInfo(BaseClientProxy *newScreen) {
 }
 
 void Server::onMouseMoveSecondary(SInt32 dx, SInt32 dy) {
-  LOG((CLOG_DEBUG2 "onMouseMoveSecondary %+d,%+d", dx, dy));
+  LOG((CLOG_DEBUG2 "onMouseMoveSecondary initial %+d,%+d", dx, dy));
+  const char *envVal = std::getenv("SYNERGY_MOUSE_ADJUSTMENT");
```
Comment (sithlord48) [inline src/lib/server/Server.cpp:1795]: Can we use `DESKFLOW_MOUSE_ADJUSTMENT` ?

**Review by sithlord48 — CHANGES_REQUESTED** [review]

**Comment by nbolton** [comment]
> > @spurin Are you happy for this to be merged in its current state?
> 
> All good by me @nbolton

@spurin I changed the env var to `DESKFLOW_MOUSE_ADJUSTMENT`. Would you mind double-checking it works as expected with the renamed env var?

**Comment by sithlord48** [comment]
Approval once @spurin  confirms the name change was successful.

**Comment by spurin** [comment]
It's looking good! Testing successful.

Screenshot 2024-09-24 at 15 12 44

**Review by nbolton — APPROVED** [review]

**Comment by nbolton** [comment]
> Approved once @spurin confirms the name change was successful.

@sithlord48 The change request is blocking the merge. Happy for me to dismiss your review, or would you prefer to re-approve?

image

**Review by sithlord48 — APPROVED** [review]

**Pull request merged**